keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,354 @@
+---
+kep-number: 35
+title: Inline CSI Volumes Support
+authors:
+  - "@vladimirvivien"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+reviewers:
+  - "@thockin"
+  - "@saad-ali"
+  - "@msau42"
+  - "@jsafrane"
+  - "@liggit"
+approvers:
+  - TBD
+creation-date: 2019-01-22
+status: implementable
+---
+
+# Inline CSI volumes support
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [User Stories](#user-stories)
+    * [Examples](#examples)
+* [Proposal](#proposal)
+    * [VolumeSource API](#volumesource-api)
+    * [Secret references](#secret-references)
+    * [Specifying allowed inline drivers with PodSecurityPolicy](#specifying-allowed-inline-drivers-with-podSecuritypolicy)
+    * [Persistent vs ephemeral lifecycle setting](#persistent-vs-ephemeral-lifecycle-setting)
+    * [VolumeHandle generation](#volumehandle-generation)
+    * [Inline CSI volume operation stages](#inline-csi-volume-operation-stages)
+    * [Security considerations](#security-considerations)
+
+## Summary
+Currently, volumes that are backed by CSI drivers can only be used with the `PersistentVolume` and `PersistentVolumeClaim` objects. This proposal is to implement support for the ability to nest CSI volume declarations within pod specs, as is supported already by existing in-tree storage providers.
+
+This KEP started life as [feature #2273](https://github.com/kubernetes/community/pull/2273).  Please follow that link for historical context.
+
+
+## Motivation
+Implementing support for embedding volumes directly in pod specs would allow external CSI drivers to have feature parity with their in-tree counter parts.  There are several reasons why this is desirable:
+* It provides the community a path to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. 
```

Comment:
Is there a tracking issue, or reference, that outlines the issues with moving * out of tree?

The grand promise of CSI is ~= CNI, it's a post bootstrap addition, and this KEP is a step in that direction.  I guess I'm looking for a roadmap or umbrella issue that outlines this.

Reply:
I don't know of a complete issue/roadmap.

I do know about this one:
https://github.com/kubernetes/community/pull/2199

Reply:
@timothysc CSI has similar aspiration of moving in-tree drivers out via CSI drivers. There is a separate set of  tasks and roadmap ([here]( https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md)) with design to track the exodus.